lib/req/steps.ex

```diff
@@ -307,6 +307,52 @@ defmodule Req.Steps do
     end
   end
 
+  @doc """
+  Use a templated request path.
+
+  ## Request Options
+
+    * `:path_params` - params to add to the templated path. Defaults to `[]`.
+
+  ## Examples
+
+      iex> Req.get!("https://httpbin.org/status/:code", path_params: [code: 200]).status
+      200
+
+  """
+  @doc step: :request
+  @rx ~r/:([a-zA-Z]{1}[\w_]*)/
+
+  def put_path_params(request) do
+    put_path_params(request, Map.get(request.options, :path_params, []))
+  end
+
+  defp put_path_params(request, []) do
+    request
+  end
+
+  defp put_path_params(request, params) do
+    request
+    |> Req.Request.put_private(:path_params_template, request.url.path)
+    |> apply_path_params(params)
+  end
+
+  defp apply_path_params(request, params) do
+    url = build_url(request.url.path, params)
+
+    uri = Map.put(request.url, :path, url)
+
+    Map.put(request, :url, uri)
+  end
+
+  defp build_url(url, nil), do: url
+
+  defp build_url(url, params) do
+    Regex.replace(@rx, url, fn match, key ->
+      to_string(params[String.to_existing_atom(key)] || match)
+    end)
+  end
```

Comment:
```suggestion
  defp apply_path_params(request, params) do
    update_in(request.url.path, fn path ->
      Enum.reduce(params, path, {name, value}, acc ->
        String.replace(acc, ~r/:#{name}(\/|$)/, "#{value}\\1")
      end)
    end)
  end
```

wdyt? I think the advantage is avoiding string-to-atom conversion.

Reply:
I think that's fine as long as we aren't using more memory by doing the reduce than the original created. Would this not create a new binary per param? I don't know enough about the internals of the regex module but did see in the source `String.replace/3` with a regex as the second arg just calls `Regex.replace`.

I guess, is it cheaper to call `to_existing_atom/1` than the cost of creating a regex struct per param?

Reply:
I thought the algorithm was simpler conceptually but it probably isn’t and what you already have is most likely faster anyway. Sorry for noise. Thanks for the PR!

Reply:
No worries. I took most of the implementation from Tesla, so hopefully they thought it through. 😂 

Thanks!

Reply:
oh, haha! I took a stab at simplifying it after all, https://github.com/wojtekmach/req/commit/b5b825d2b7641d7a393d67044782c3f87aa43013, and now I _really_ hope I didn't screw anything up!

Reply:
Heh. Looks good to me!

Reply:
Oh. And a callout in case this gets touched again in the future, [this line](https://github.com/wojtekmach/req/commit/b5b825d2b7641d7a393d67044782c3f87aa43013#diff-07dc6fd4ec95415d610a3e3ddf9374540d6d64d02eaf42df7a233d6f05b32cc1R334) needs to be there so otel can pick out the original templated path for the span name. It doesn't have a direct purpose in the step itself, but allows a backdoor access to that original value.

Reply:
the gh link didn't work for me but I assume you meant this one?

```elixir
|> Req.Request.put_private(:path_params_template, request.url.path)
```